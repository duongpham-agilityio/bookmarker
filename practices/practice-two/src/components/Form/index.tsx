import { MouseEvent, memo } from 'react';

// Hooks
import { useForm } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';

// Types
import { Book } from 'types';

// Styles
import styles from 'components/Form/index.module.css';

// Assets
import UploadIcon from 'assets/icons/upload.svg';

export type FormProps = {
  value: Omit<Book, 'publishDate' | 'deletedAt' | 'createdAt' | 'updatedAt'> & {
    publishDate?: number;
  };
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: (_event: MouseEvent) => void;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit,
  } = useForm(data, type);

  return (
    <section
      className={styles.overlay}
      onClick={(e: MouseEvent) => {
        if (onClose) onClose(e);
      }}
    >
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmit}
        onClick={(e: MouseEvent) => {
          e.stopPropagation();

          resetRecommended([]);
        }}
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <ul className={styles.recommend}>
                  {booksRecommended.map((book, index) => {
                    return (
                      <li
                        className={styles.recommendItem}
                        key={index}
                        onClick={(event: MouseEvent) => {
                          event.preventDefault();

                          handleSelectRecommended(index);
                        }}
                      >
                        <p>{book.name}</p>
                      </li>
                    );
                  })}
                </ul>
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />
                {!imageName ? (
                  <>
                    <Button
                      label="Upload"
                      width="w-lg"
                      border="b-lg"
                      variant="primary"
                      leftIcon={UploadIcon}
                      onClick={() => {
                        refImage.current?.click();
                      }}
                    />
                  </>
                ) : (
                  <p
                    onClick={() => {
                      refImage.current?.click();
                    }}
                  >
                    {imageName}
                  </p>
                )}
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
            />
            <Button
              disabled={isUpload}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form);
