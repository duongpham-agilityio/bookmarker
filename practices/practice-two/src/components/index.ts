export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Form, type FormProps } from './Form';
export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
