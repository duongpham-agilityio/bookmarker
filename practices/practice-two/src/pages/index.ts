import { lazy } from 'react';

const Home = lazy(() => import('pages/Home'));
const Detail = lazy(() => import('pages/Detail'));

export { Home, Detail };
