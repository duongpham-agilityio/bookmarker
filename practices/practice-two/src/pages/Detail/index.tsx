import { useContext, useMemo } from 'react';
import { useNavigate } from 'react-router-dom';
import { useBook } from 'hooks';

// HOCs
import { withErrorBoundaries } from 'hocs/withErrorBoundaries';

// Contexts
import { FormContext } from 'contexts/Form/context';

// Components
import { Button, Heading } from 'components/commons';
import { Error } from 'components';
import DetailSkeleton from 'pages/Detail/Skeleton';

// Styles
import containerStyles from 'styles/commons/index.module.css';
import styles from 'pages/Detail/index.module.css';

// Assets
import TrashIcon from 'assets/icons/trash.svg';
import BackIcon from 'assets/icons/back.svg';
import PenCilIcon from 'assets/icons/pencil.svg';
import { convertDateTimeToTimeString, convertTimeToDate } from 'helpers';

const Detail = () => {
  const { data, error, isLoading, deleteBook } = useBook();
  const redirect = useNavigate();

  if (error) {
    return <Error />;
  }

  const { dispatch } = useContext(FormContext);
  const book = useMemo(() => {
    const { publishDate, createdAt, updatedAt, ...rest } = data;
    const publishDateConvert = `${convertDateTimeToTimeString(
      publishDate
    )}, ${convertTimeToDate(publishDate, '/', true)}`;
    const createdAtConvert = `${convertDateTimeToTimeString(
      createdAt
    )}, ${convertTimeToDate(createdAt, '/', true)}`;
    const updatedAtConvert = `${convertDateTimeToTimeString(
      updatedAt
    )}, ${convertTimeToDate(updatedAt, '/', true)}`;

    return {
      ...rest,
      publishDate: publishDateConvert,
      createdAt: createdAtConvert,
      updatedAt: updatedAtConvert,
    };
  }, [data]);

  return (
    <main className={containerStyles.container}>
      {isLoading ? (
        <DetailSkeleton />
      ) : (
        <section className={styles.detail}>
          <div className={styles.detailItem}>
            <Heading label={data.name} size="xl" className={styles.heading} />
            <div>
              <div className={styles.action}>
                <Button
                  label=""
                  leftIcon={TrashIcon}
                  variant="danger"
                  onClick={deleteBook}
                />
                <Button
                  label=""
                  leftIcon={BackIcon}
                  variant="primary"
                  onClick={() => redirect('/books')}
                />
                <Button
                  label="Edit"
                  leftIcon={PenCilIcon}
                  variant="primary"
                  width="w-lg"
                  onClick={() =>
                    dispatch({
                      formData: data,
                      title: 'Edit book',
                      type: 'update',
                    })
                  }
                />
              </div>
            </div>
          </div>
          <div className={styles.detailItem}>
            <div className={styles.info}>
              <p className={styles.description}>
                <span>{data.description.substring(0, 150)}</span>
                <span className={styles.actionText}>show more</span>
              </p>
              <ul className={styles.listInfo}>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Author: </span>
                  <span className={styles.text}>{book.author}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Published date: </span>
                  <span className={styles.text}>{book.publishDate}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Created at: </span>
                  <span className={styles.text}>{book.createdAt}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Updated at: </span>
                  <span className={styles.text}>{book.updatedAt}</span>
                </li>
              </ul>
            </div>
            <div className={styles.imageWrap}>
              <img
                src={data.imageURL}
                alt={data.name}
                className={styles.image}
              />
            </div>
          </div>
        </section>
      )}
    </main>
  );
};

export default withErrorBoundaries(Detail);
