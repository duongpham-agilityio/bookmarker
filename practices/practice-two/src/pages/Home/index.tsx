import { ChangeEvent, MouseEvent, useContext, useState } from 'react';
import { Link } from 'react-router-dom';
import { useBooks, useDebounce } from 'hooks';

// HOCS
import { withErrorBoundaries } from 'hocs/withErrorBoundaries';

// Contexts
import { FormContext } from 'contexts/Form/context';

//Components
import { Button, Heading, Input } from 'components/commons';
import { Card, CardSkeleton, Error } from 'components';

// Constants
import { ENDPOINT, MESSAGES, SORT, TITLE_FORM } from '@constants';

// Styles
import homeStyles from 'pages/Home/index.module.css';
import commonStyles from 'styles/commons/index.module.css';

// Assets
import SearchIcon from 'assets/icons/search.svg';
import AddIcon from 'assets/icons/add.svg';

// Helpers
import { convertDateTimeToTimeString } from 'helpers';

const Home = () => {
  const { dispatch } = useContext(FormContext);
  const {
    param: { page: currentPage, sort },
    data: dataShow,
    isLoading,
    error,
    pagination,
    deleteBook,
    setSearchParam,
    convertSearchParamsToString,
  } = useBooks();
  const [search, setSearch] = useState('');
  const debounce = useDebounce((value) => setSearchParam('name', value));

  if (error) {
    return <Error />;
  }

  return (
    <main className={commonStyles.container}>
      <section className={homeStyles.home}>
        <div className={homeStyles.navbar}>
          <Input
            className={homeStyles.search}
            value={search}
            leftIcon={SearchIcon}
            placeholder="Search something..."
            onChange={(event: ChangeEvent<HTMLInputElement>) => {
              setSearch(event.target.value);
              debounce(event.target.value);
            }}
          />
          <nav className={homeStyles.navigation}>
            <ul className={homeStyles.navList}>
              <li>
                <Link
                  to={`/${ENDPOINT.BOOKS}${convertSearchParamsToString(
                    'sort',
                    SORT.ASCENDING
                  )}`}
                  className={`${homeStyles.navLink} ${
                    sort === SORT.ASCENDING ? homeStyles.active : ''
                  }`}
                >
                  Ascending
                </Link>
              </li>
              <li>
                <Link
                  to={`/${ENDPOINT.BOOKS}${convertSearchParamsToString(
                    'sort',
                    SORT.DESCENDING
                  )}`}
                  className={`${homeStyles.navLink} ${
                    sort === SORT.DESCENDING ? homeStyles.active : ''
                  }`}
                >
                  Descending
                </Link>
              </li>
            </ul>
          </nav>
          <Button
            label="Create"
            variant="primary"
            leftIcon={AddIcon}
            width="w-lg"
            border="b-lg"
            onClick={() => {
              dispatch({
                formData: {
                  author: '',
                  description: '',
                  imageURL: '',
                  name: '',
                  createdAt: new Date().getTime(),
                  deletedAt: null,
                  updatedAt: new Date().getTime(),
                },
                title: TITLE_FORM.CREATE,
                type: 'create',
              });
            }}
          />
        </div>

        {isLoading ? (
          <div className={homeStyles.grid}>
            {[1, 2, 3, 4, 5, 6].map((value) => (
              <CardSkeleton key={value} />
            ))}
          </div>
        ) : (
          <>
            {dataShow?.length ? (
              <>
                <div className={homeStyles.content}>
                  <div className={homeStyles.grid}>
                    {dataShow.map((book) => (
                      <Card
                        href={`/${ENDPOINT.BOOKS}/${book.id}`}
                        title={book.name}
                        description={book.description}
                        publishedDate={convertDateTimeToTimeString(
                          book.createdAt
                        )}
                        imageUrl={book.imageURL}
                        key={book.id}
                        onDelete={(event: MouseEvent) => {
                          event.preventDefault();

                          deleteBook(book.id || 0);
                        }}
                      />
                    ))}
                  </div>
                </div>
                <div className={homeStyles.pagination}>
                  {pagination.length > 1 &&
                    pagination.map((__, index) => {
                      const page = index + 1;

                      return (
                        <Button
                          label={`${page}`}
                          variant={
                            currentPage === page ? 'primary' : 'secondary'
                          }
                          size="small"
                          key={index}
                          onClick={() => {
                            setSearchParam('page', `${page}`);
                          }}
                        />
                      );
                    })}
                </div>
              </>
            ) : (
              <div className={homeStyles.empty}>
                <Heading
                  label={MESSAGES.EMPTY_TITLE}
                  className={homeStyles.emptyTitle}
                  size="xl"
                />
                <p className={homeStyles.emptyDescription}>
                  {MESSAGES.EMPTY_DESCRIPTION}
                </p>
              </div>
            )}
          </>
        )}
      </section>
    </main>
  );
};

export default withErrorBoundaries(Home);
