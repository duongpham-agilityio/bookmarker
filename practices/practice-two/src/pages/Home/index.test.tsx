import { act, render } from '@testing-library/react';
import { CardProps } from 'components';

// Component
import { Home } from 'pages';
import { Suspense } from 'react';

const setup = async () =>
  await act(() =>
    render(
      <Suspense fallback={<p>Loading...</p>}>
        <Home />
      </Suspense>
    )
  );

jest.mock('components', () => ({
  Card: (props: CardProps) => <p>{props.title}</p>,
}));

jest.mock('assets/icons/search.svg', () => ({
  default: 'assets/icons/search.svg',
}));

jest.mock('assets/icons/add.svg', () => ({
  default: 'assets/icons/add.svg',
}));

describe('Home', () => {
  it('match snapshot', async () => {
    const { asFragment } = await setup();
  });
});
