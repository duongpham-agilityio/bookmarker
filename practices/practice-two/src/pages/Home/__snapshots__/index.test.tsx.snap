// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home match snapshot 1`] = `
<DocumentFragment>
  <main>
    <section>
      <div>
        <div
          class="undefined function search() { [native code] }"
        >
          <img
            alt="icon"
            src="assets/icons/search.svg"
          />
          <input
            placeholder="Search something..."
            type="text"
            value=""
          />
        </div>
        <nav>
          <ul>
            <li>
              <a
                class="undefined "
                href="/?sort=ascending"
              >
                Ascending
              </a>
            </li>
            <li>
              <a
                class="undefined "
                href="/?sort=descending"
              >
                Descending
              </a>
            </li>
          </ul>
        </nav>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          <img
            alt="icon action"
            src="assets/icons/add.svg"
          />
          Create
        </button>
      </div>
      <div>
        <div>
          <p>
            HTML/CSS Ebook 1
          </p>
          <p>
            HTML/CSS Ebook 2
          </p>
          <p>
            HTML/CSS Ebook 3
          </p>
        </div>
      </div>
      <div>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          1
        </button>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`Home render 1`] = `<DocumentFragment />`;
